src/dataframe/dataframe.jl

```diff
@@ -303,11 +276,16 @@ function Base.getindex(df::DataFrame, row_inds::AbstractVector, col_inds::Abstra
 end
 
 # df[:, SingleColumnIndex] => AbstractVector
+function Base.getindex(df::DataFrame, row_inds::Colon, col_ind::ColumnIndex)
+    selected_column = index(df)[col_ind]
+    copy(_columns(df)[selected_column][row_inds])
+end
+
 # df[:, MultiColumnIndex] => DataFrame
-function Base.getindex(df::DataFrame, row_ind::Colon, col_inds)
-    Base.depwarn("Indexing with colon as row will create a copy in the future. " *
-                 "Use `df[col_inds]` to get the columns without copying", :getindex)
-    df[col_inds]
+function Base.getindex(df::DataFrame, row_ind::Colon, col_inds::AbstractVector)
+    selected_columns = index(df)[col_inds]
+    new_columns = AbstractVector[copy(dv) for dv in _columns(df)[selected_columns]]
```

Comment:
Maybe it would be worth avoiding two allocations by iterating over `col_inds` and calling `_columns(df)[index(df)[col]]` on the fly in the comprehension? Not sure.

Reply:
This will not be efficient when we pass `col_inds` as symbols because we would have to do the lookup twice.
The alternative is to have three separate implementations for indexing with `Bool`, integers and `Symbols` - it is possible to improve the performance then, but I would leave it for a separate PR as it is significantly complex (what we do now replicates what was used in the past and is safe as we delegate handling of different corner cases to `Index`, where I made sure that all is "correct" - hopefully 😄).